### ISSUE TITLE
Install action does not install anything if ran previously with ClientOnly flagged

### ISSUE THREAD

**Issue by delca85** [helm/helm#11463]
With the current Action Go package APIs, I am having an issue with the [install action](https://github.com/helm/helm/blob/f08db5c899c4e312c72ae863acbb90ac8a4747ed/pkg/action/install.go#L487).

The issue I am facing is that if I run two install actions, the first with `ClientOnly` flag set to `true`, also the second one does not reach the cluster.

Here's the code of the 1st action that runs to produce template files in the `outputDir`:
```go
client := action.NewInstall(settings)
client.DryRun = true
client.ReleaseName = "my-test-release"
client.Replace = true
client.ClientOnly = true
client.OutputDir = outputDir
client.UseReleaseName = true
_, err = client.RunWithContext(ctx, chart, values)
```

Then, when the flow proceeds, I run another install action with the aim of actually installing the release:
```go
client := action.NewInstall(settings)
client.ReleaseName = releaseName
client.Namespace = "my-test-namespace"
client.DryRun = false
client.ClientOnly = false
client.Wait, client.WaitForJobs, client.Timeout = release.GetWaitConditions()
rel, err = client.RunWithContext(ctx, chart, values)
```

The second action does not produce any error but it actually does not deploy anything.
If, in exactly the same flow, the first install action runs with `ClientOnly = false`, the second install works as expected and actually installs the release into the cluster.

**Comment by delca85** [comment]
From my understanding, the issue is [this line](https://github.com/helm/helm/blob/d79ae9f927c00069ec39571501e82a9a8d0697ef/pkg/action/install.go#L224) that assign to the shared `KubeClient` a fake one.

The workaround I came up with is to reinitialize it:
```go
d.helmActionConfig.KubeClient = kube.New(newRESTClientGetter(d.Namespace, d.helmCliSettings))
```

This line makes the install action able to install the release.

I'd like to know if this is the expected behaviour.

**Comment by joejulian** [comment]
IMHO, I'd say the issue is at https://github.com/helm/helm/blob/d79ae9f927c00069ec39571501e82a9a8d0697ef/pkg/action/install.go#L127-L134

I would expect that a "New" anything should be creating a completely new independent resource. The cfg shouldn't just be a link to the configuration that was passed to it, it should be a copy.

I'm going to triage this as a bug and see if anybody disagrees.

**Comment by joejulian** [comment]
A workaround, of course, is to pass NewInstall a copy of the Configuration struct.

If I were to write a PR to address this, I think I would add a DeepCopy method for the Configuration struct and then use it in NewInstall for setting cfg.

**Comment by delca85** [comment]
> https://github.com/helm/helm/issues/11463#issuecomment-1289808607

Yes, I totally agree. What I meant in the issue description is that the behaviour was due to that line, that makes totally sense in order to perform a mock call, but shouldn't be spread across all the actions created after that one.

@joejulian May I take this issue?

**Comment by delca85** [comment]
@joejulian I am looking into this and I am wondering about the approach to follow, to be as more compliant as possible with the project. I think there are several paths that might be followed:

1. to use an external library (e.g. [https://github.com/mohae/deepcopy](https://github.com/mohae/deepcopy), [https://github.com/ulule/deepcopier](https://github.com/ulule/deepcopier), [github.com/mitchellh/copystructure](https://github.com/mitchellh/copystructure))
2. to generate the DeepCopy method through, for example, [deepcopy-gen](https://pkg.go.dev/k8s.io/gengo/examples/deepcopy-gen)
3. to implement a method that deeply copy the `Configuration` struct, manually copying its fields recursively

I think I'd go with the third option, to avoid to add dependencies that have not been needed till now but I'd like to know your thoughts about before starting the implementation. Thanks!

**Comment by joejulian** [comment]
I lean toward the third option but that does create a (small) maintenance burden and a pretty easy failure path should the struct be updated but the contributor doesn't realize they need to change the DeepCopy.

Mitch's library is likely to remain stable and maintained, and has a compatible license.

If I were going to use a generator, I'd use [controller-gen](https://github.com/kubernetes-sigs/controller-tools). We would only need the +object generator that generates code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.

@helm/helm-core-maintainers should probably weigh in on this.

**Comment by delca85** [comment]
Yes, it would be great to receive some input from anyone within @helm team. @yxxhero , can you help here? Thanks!

**Comment by delca85** [comment]
@yxxhero I am sorry 😞 
I'd be more than happy to raise a PR to solve this issue, I am only aiming to receive some input about the best approach to follow to fix this.

**Comment by delca85** [comment]
@mattfarina @sabre1041 may you provide any help here? Thanks!

**Comment by joejulian** [comment]
@delca85 I'd say just go ahead and create a PR with the simplest solution.

**Comment by delca85** [comment]
@joejulian, thanks I'll definitely do it! May you assign this to me?

**Comment by adityaraj178** [comment]
Hi @delca85 ,
I noticed this issue is still open and I'd like to contribute. Are you still planning on working on it?
If you are no longer working on it, may I take this over and submit a pull request?

Thank you!
